### ISSUE TITLE
Version 2.0

### ISSUE THREAD

**Issue by 1271** [yuru-yuri/manga-dl#64]

**Comment by gromenauer** [comment]
> If there are any wishes, I'll be happy to hear it.

* Colored exceptions (`import [better_exceptions](https://github.com/qix-/better-exceptions)`)
* Colored logs (`import [colorlog](https://pypi.org/project/colorlog/)` or [zenlog](https://github.com/manufacturaind/python-zenlog/)?)
  * Volume number in the logs
* Command line tab completion (`import [argcomplete](https://pypi.org/project/argcomplete/)`)
* A Dockerfile (maybe using https://hub.docker.com/_/python/). Setup and bug reporting will be easy.
* Think a way to know what was downloaded in the past, even if you deleted it or moved to other disk (to have disk space). (csv, yaml, sqlite, ... what could be best?).

This is what I can think of now.

**Comment by phanirithvij** [comment]
Then I bet this will exceed tachyomi (even though it already is. As 250 sources is incredible as of now.)

**Comment by phanirithvij** [comment]
> Command line tab completion (import [argcomplete](https://pypi.org/project/argcomplete/))

Definitely.

**Comment by phanirithvij** [comment]
- ```info.json``` along with or instead of ```info.txt```

- Include website name in Manga directory while downloading the manga (non conflicting directory names for same manga from different sources)
ex:
```
Manga/Berserk-mangarock_com/
Manga/Berserk-kissmanga_com/
```
Or
```
Manga/mangarock_com/Berserk
Manga/kissmanga_com/Berserk
```

- ```-z or --zip``` only if user wants output as zip files else just store images in directories

- ```--html``` to generate html in each directory (of volumes like ```vol-000.html```)
 (for users to read chapters in browser) with a nav on top (source > manga > chapter) and
with some css to make it look like mangarock (or something cooler)

I will add more if you think at least one of the above seems _OK_

**Comment by 1271** [comment]
Thank you. I'll definitely take this into consideration!

---

> Colored

It is not difficult.

> Command line tab completion

It will be wonderful. I've been thinking about this for a long time. I did not know about this plugin. Thank you.

> A Dockerfile

For most, this will be more difficult than running the pip install command.
But I'll think about it. (It already was, but in a consequence I deleted the docker file as unnecessary)

> Think a way to know what was downloaded in the past.

Such a thought was. I think it can be done.
The truth does not see practical meaning. It will be only history.
If you moved files to another location, manga-py will not be able to find out about it.
Although if you yourself pointed the folder, then it makes sense.
On the other hand, if you specified a folder, what kind of behavior should manga-py have if you are in a different folder and load manga with the default setting?
Pumping what already exists? But what if you want to download it again?

---

> info.json

It's bound to be. Perhaps even this will not be stored in the folder, but in the database (see the comment above).
The database may well be a good solution to the problem of --update-all and storing information about manga as a whole.
But if you need a separate file for each manga, it can also be arranged.

> -z or --zip

OK. I'll add this option.
It's just that I use my phone for reading. A program for reading works fine with archives. However, the manga pages do not fit into the phone's image list. Which is very convenient.

> --html

A good idea! This will also be done!

>I will add more if you think at least one of the above seems OK

Yes, I would like to hear as many wishes as possible, in order to change the structure of manga-py as little as possible afterwards. this entails a lot of "crutches". What I would like to avoid.

**Comment by phanirithvij** [comment]
#40 add an API

**Comment by 1271** [comment]
Sure. I have not forgotten about this.

**Comment by 1271** [comment]
> Include website name in Manga directory

Usually it does not make sense. But you can make it optional

**Comment by phanirithvij2000** [comment]
New source [Zmanga Spanish](http://zmanga.net)

**Comment by 1271** [comment]
> New source Zmanga Spanish

Thank you.
But, please, next time create a separate issue. It's easier to follow you and me.

**Comment by gromenauer** [comment]
If you want to learn more python you could use:
* AsyncIO. Instead of threads, better when th e bottleneck is I/O
* Microservices. Put a microservice as backend and you could make separate front-ends (cli, web, third party clients), easier test, deploy to docker, ...
(Maybe with [hug](https://github.com/timothycrosley/hug))

**Comment by 1271** [comment]
> AsyncIO.

Thanks. But threads work fine. Because the most narrow place in manga-py is the Internet connection.

> Microservices

It does not differ in different languages. But I do not want to write too bulky api, or do something like this. This goes beyond the scope of the project.


The new version will have a fairly flexible structure. Later manga-py will be connected as a package to the project gui and web.
In fact, there will only need to call one class.
Also I will make examples showing the possibility of connecting their providers.

Thank you for participating. I'm happy that people are interested in the project.

**Comment by gromenauer** [comment]
Have you tried...?:
* [pathlib](https://docs.python.org/3/library/pathlib.html). For me, far better than os.path.
* [requests-html](https://github.com/kennethreitz/requests-html). Seems pretty handy.

**Comment by 1271** [comment]
> pathlib

I'm more used to os.path

> requests-html

This is just a wrapper over lxml.
Maybe it's more convenient, but I'm more used to lxml.

**Comment by 1271** [comment]
@phanirithvij2000 
Check version 1.3.5, please.

**Comment by gromenauer** [comment]
Just another idea for 2.0
If you want to store data in a database think on storing the md5 for each downloaded page. Why?:
* If two chapters have the same md5 for all of each pages, then is a duplicate.
* If one page is the same for all the chapters in a comic, then is a cover page.
* If one page is the same for may comic, then is a scanlation group cover page (or something similar).

If the contents of the database is uploadable, then you can discover automatically:
* Renamed chapters
* Equivalent comics between webs
* ...

**Comment by 1271** [comment]
Hah)
I'll think about it.
But I do not think that this will be implemented. I do not want to promise anything.

**Comment by wilfriedroset** [comment]
I've been using for a moment [manga_downloader](https://github.com/joaquinpf/manga_downloader) and I really like the json library feature.
This basically allow you to define your library in a json file (with source website and destination folder). Once you have done that you simply need to execute the tool periodically with a crontab for example.

Example
```
{
    "manga_series": [
        {
            "download_path": "/mnt/manga/hunter-x-hunter/",
            "host_site": "MangaReader",
            "last_chapter_downloaded": "hunter x hunter 390",
            "name": "hunter-x-hunter",
            "timestamp": "2018-12-28 15:21:58"
        },
--8<--
```

I feel like manga-dl is missing a configuration file

**Comment by 1271** [comment]
Thank.

I thought about the data storage in json, but at the end I stopped at the option with sqlite.

In manga-py 2, an update of the already loaded manga will be implemented. (Checking new chapters)

**Comment by 1271** [comment]
docker is not particularly necessary for users of * nix, and windows users do not think that they will want to install it for themselves.
(sorry if it is not clear)

**Comment by gromenauer** [comment]
If you want to search for ideas, take a look in https://github.com/dipu-bd/lightnovel-crawler
Is other type of crawler, but have curious features.

**Comment by hanetzer** [comment]
If you're still looking for ideas, another viz.com nit; it seems they time you out eventually when downloading.
Perhaps adding a chapter/page delay option in would be useful. Letting a 'slow' process continue on in the
background is a lot better than babysitting a fast process that needs to be restarted from time to time.

**Comment by 1271** [comment]
Hi. I do not fully understand your idea.
Yes, viz.com images have an expiration time.

**Comment by hanetzer** [comment]
I mean, manga-py accesses the pages far faster than a normal human can, and
eventually it stops downloading after 10-20 volumes and begins spamming
```[MissingSchema("Invalid URL 'Retry later': No schema supplied. Perhaps you meant http://Retry later?")]```
Sometimes the last chapter downloaded appears to be corrupted as well. I think
maybe if there were a time delay between download of each page it may be able
to continue longer. I have a paid account and just hate their UI, its basically impossible
to use for extended periods of time so its easier to download the pages and use an
external viewer.

Also if you need someone to test viz stuff I can do it. I'm paid, and live in the US, so
should be no issues on my end.

**Comment by 1271** [comment]
Oh, you're right about the speed. In general, this is not difficult to do, yes. I'll add a new argument today.

Also I would be grateful to you for checking. Thanks

After my work (16:00 UTC) I will add this to the active branch. 

Something like `--wait-after-page` and` --wait-after-chapter`

**Comment by hanetzer** [comment]
Many thanks, very responsive. 10/10 would fork again :)

**Comment by 1271** [comment]
Please, check new version (1.29.0)
You can also consider reducing the number of simultaneously downloaded files (--max-threads) // default: 4

**Comment by hanetzer** [comment]
yeah I was already setting it to -t 1 as a hopeful fix to my observed issues. On another note,
it would maybe be useful to get the metadata.json saved per volume as well, useful for organizing
and such without having to manually open the cbz files (unless I misunderstand the save-chapter-info
flag option)

**Comment by 1271** [comment]
Yes, the information is now stored directly in the archive.
This may be changed someday, but not in the 1.x version.
